rts/Rendering/GL/State.h

```diff
@@ -0,0 +1,180 @@
+/* This file is part of the Recoil engine (GPL v2 or later), see LICENSE.html */
+
+#pragma once
+
+#include "myGL.h"
+#include "glHelpers.h"
+#include "System/FuncUtil.h"
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+/*
+	GL::State doesn't represent active GL state.
+	It represents GL state, that was last applied by GL::SubState objects, or any other object that performs state change via this interface.
+
+	If a state was not set by a SubState object, it is internally set to Unknown.
+	When a state is Unknown and SubState object requests it, the state is fetched from GL.
+	When the last existing SubState object reverts a state, it is set back to Unknown.
+	Setting back to Unknown after the last SubState object enforces fetch from GL every time the first SubState object is created,
+		but makes it possible to insert SubState object in any part of the code without serious restrictions to the whole prior code.
+
+	User must make sure that ordinary GL commands that are not tracked do not collide with SubState object's modifications during its lifetime.
+*/
+
+namespace GL
+{
+
+template<auto DedicatedGLFuncPtrPtr, GLenum... GLParamNames> class StateAttribute {
+private:
+	using GLParamsType = FuncPtrSignature<DedicatedGLFuncPtrPtr, GLboolean>;
+	using StatusType = bool;
+	static constexpr StatusType Unknown = false;
+	static constexpr StatusType WasUnknown = false;
+
+public:
+	using ValueType = std::pair<StatusType, GLParamsType>;
```

Comment:
No option to use `std::optional<GLParamsType>` ?

Reply:
No, the values must be used even when false. It is not optional by design.